README.rst
oauth_redirect
==============

Securely redirects OAuth responses to known clients.

Description
-----------

During the OAuth workflow a webserver defined by redirect_uri receives the
result of a the authorization given by the enduser.

This redirect_uri is usually a whitelisted uri to avoid phising attacks.

Unfortunately this makes it hard to write an OAuth integrated client to an
API where the domain is not known in advance.

Faced with such a situation we wrote this server to move the decision of
who can receive the result of an OAuth handshake from the OAuth provider
to an intermediary.

How it Works
------------

Using oauth_redirect the OAuth workflow works as follows:

1. The oauth_redirect server is run on a TLS protected site. For example:
   https://oauth.seantis.ch.

2. The OAuth provider is configured to allow redirects to
   https://oauth.seantis.ch/redirect.

3. The client that wants to acquire the authorisation registers itself with
   the oauth_redirect server using a secret authentication code.

4. The enduser is presented with the OAuth authorization site, with the
   redirect_uri set to https://oauth.seantis.ch/redirect.

5. The result is sent to the oauth_redirect server, which will forward/proxy
   the request to the client, if and only if the client has registered itself
   before and the request from the OAuth provider contains a token to that
   effect.

Methods
-------

``POST /register/<authentication code>``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Used by the client/authorisation seeker to register itself. The client is
required to include a secret authentication code in the registration message.

The body of the POST request is a json in this form::

    {
        'url': "The url that handles the forwarded OAuth response.",
        'method': "The method with which the url should be called (get, put or post).",
        'success_url': "The url the server redirects to if the handler url returns a 2XX code.",
        'error_url': "The url the server redirects to if the handler url returns a non-2XX code.",
        'ttl': "The optional time to live in seconds (defaults to 3600 seconds)",
        'secret': "A client-specific secret that should be used authenticate
        the forwarded request. If the request does not contain this secret,
        someone other than oauth_redirect has sent it.",
    }

Returns the token which needs to be passed by the OAuth provider::

    {
        'token': "..."
    }

``(GET|POST) /redirect``
~~~~~~~~~~~~~~~~~~~~~~~~

The endpoint communicated to the OAuth provider through the redirect_uri. To
authenticate the request coming from the OAuth provider must contain the
token given by ``/register/<authentication code>``.

Usually OAuth providers provide some kind of of value that may be passed from
the client to the redirect_uri. This value can be used to carry the token
back to the oauth_redirect server.

If there is no such value, the token may also be passed by url, using query
paramters (i.e. https://oauth.seantis.ch/redirect?token=...).

Any value will do, a value in a json body, a formdata value or a query
parameter.

If the redirect request is accepted it is proxied to the registered url. The
result of the ``/redirect`` request is the result of the proxied url.

If the request was accepted, it is deleted.

Deployment
----------

The server is implemented using `aiohttp <http://aiohttp.readthedocs.io/en/stable/>`_.
It requires at least Python 3.5.

Though it might be possible to implement TLS support on the oauth_redirect we
recommend that you put it behind a proper web proxy like nginx/apache.

To run the server run::

    oauth-redirect --host localhost --port 8080 --database registered --auth <your custom auth code>


Run the Tests
-------------

Install tox and run it::

    pip install tox
    tox

Limit the tests to a specific python version::

    tox -e py27

Conventions
-----------

Oauth_redirect follows PEP8 as close as possible. To test for it run::

    tox -e pep8

Oauth_redirect uses `Semantic Versioning <http://semver.org/>`_

Build Status
------------

.. image:: https://travis-ci.org/seantis/oauth_redirect.png
  :target: https://travis-ci.org/seantis/oauth_redirect
  :alt: Build Status

Coverage
--------

.. image:: https://coveralls.io/repos/seantis/oauth_redirect/badge.png?branch=master
  :target: https://coveralls.io/r/seantis/oauth_redirect?branch=master
  :alt: Project Coverage

Latest PyPI Release
-------------------

.. image:: https://badge.fury.io/py/oauth_redirect.svg
    :target: https://badge.fury.io/py/oauth_redirect
    :alt: Latest PyPI Release

License
-------
oauth_redirect is released under GPLv2


HISTORY.rst
Changelog
---------

0.3.0 (2017-05-24)
~~~~~~~~~~~~~~~~~~~~~

- Adds the ability to define a success and an error url.
  [href]

0.2.0 (2017-05-24)
~~~~~~~~~~~~~~~~~~~~~

- Adds the ability to define the HTTP method with which the redirect target
  should be called.
  [href]

- Adds support for GET in addition to POST for the redirect view handler.
  [href]

0.1.0 (2017-05-23)
~~~~~~~~~~~~~~~~~~~~~

- Initial Release.
  [href]


oauth_redirect/tests/conftest.py
import json
import morepath
import os
import pytest
import tempfile

from mirakuru import TCPExecutor
from pathlib import Path
from pytest_localserver.http import WSGIServer
from uuid import uuid4


def create_wsgi_endpoint_app(output_path):

    class App(morepath.App):
        pass

    @App.path('/')
    class Endpoint(object):
        pass

    @App.view(model=Endpoint, request_method='POST')
    def post_endpoint(self, request):
        with (output_path / 'POST.json').open('w') as f:
            f.write(request.text)

    @App.view(model=Endpoint, request_method='GET')
    def get_endpoint(self, request):
        with (output_path / 'GET.json').open('w') as f:
            f.write(json.dumps(dict(request.params)))

    @App.view(model=Endpoint, request_method='PUT')
    def put_endpoint(self, request):
        with (output_path / 'PUT.json').open('w') as f:
            f.write(request.text)

    return App()


@pytest.fixture(scope='module')
def temporary_path():
    with tempfile.TemporaryDirectory() as tempdir:
        yield Path(tempdir)


@pytest.fixture(scope='module')
def json_endpoint(temporary_path):
    server = WSGIServer(application=create_wsgi_endpoint_app(temporary_path))
    server.start()
    server.temporary_path = temporary_path

    yield server

    server.stop()


@pytest.fixture(scope='module')
def oauth_redirect_server(temporary_path):
    config = temporary_path / 'config.yml'
    database = temporary_path / 'registered_clients'
    auth = uuid4().hex

    with config.open('w') as f:
        f.write('\n'.join((
            'database: {}'.format(database),
            'auth: {}'.format(auth)
        )))

    command = 'oauth-redirect -h 127.0.0.1 -p 8000 -a {} -d {}'.format(
        auth, database
    )

    server = TCPExecutor(command, host='127.0.0.1', port=8000, shell=True)
    server.start()
    server.config = config
    server.database = database
    server.auth = auth
    server.url = 'http://127.0.0.1:8000'

    yield server

    server.stop()


@pytest.fixture(scope='function', autouse=True)
def clear_temporary_path(temporary_path):
    for f in os.scandir(str(temporary_path)):
        os.unlink(f.path)


oauth_redirect/tests/test_server.py
import json
import pytest
import requests
import time


def test_unauthorized(oauth_redirect_server):
    url = oauth_redirect_server.url

    assert requests.post('{}/register'.format(url)).status_code == 404
    assert requests.post('{}/register/foobar'.format(url)).status_code == 401


def test_unknown_oauth_request(oauth_redirect_server):
    url = '{}/redirect'.format(oauth_redirect_server.url)

    assert requests.post(url).status_code == 404
    assert requests.post(url, json={'state': '0xunknown'}).status_code == 404


def test_end_to_end(oauth_redirect_server, json_endpoint):
    register = '{}/register/{}'.format(
        oauth_redirect_server.url, oauth_redirect_server.auth)

    redirect = '{}/redirect'.format(
        oauth_redirect_server.url)

    # run the whole thing end-to-end
    response = requests.post(register, json={
        'url': json_endpoint.url,
        'secret': "Area 51",
        'method': "POST",
        'success_url': 'http://example.org',
        'error_url': 'http://example.org'
    })

    assert response.status_code == 200
    data = response.json()

    assert 'token' in data
    assert data['token'].startswith('0x')

    response = requests.post(redirect, json={
        'foo': 'bar',
        'bar': data['token']
    }, allow_redirects=False)

    assert response.status_code == 302

    with (json_endpoint.temporary_path / 'POST.json').open('r') as f:
        received = json.loads(f.read())

    assert received['foo'] == 'bar'
    assert received['bar'] == data['token']
    assert received['oauth_redirect_secret'] == "Area 51"

    # try to repeat the request using the same code
    assert requests.post(redirect, json={
        'foo': 'bar',
        'bar': data['token']
    }).status_code == 404

    # try to send a request after it's expiry has been reached
    response = requests.post(register, json={
        'url': json_endpoint.url,
        'ttl': 1,
        'secret': "Area 51",
        'method': "POST",
        'success_url': 'http://example.org',
        'error_url': 'http://example.org'
    })
    data = response.json()

    time.sleep(2)

    assert requests.post(redirect, json={
        'foo': 'bar',
        'bar': data['token']
    }).status_code == 404


@pytest.mark.parametrize('method', [
    'GET',
    'PUT',
    'POST'
])
def test_request_methods(oauth_redirect_server, json_endpoint, method):
    register = '{}/register/{}'.format(
        oauth_redirect_server.url, oauth_redirect_server.auth)

    redirect = '{}/redirect'.format(
        oauth_redirect_server.url)

    # simulate oauth provider that sends post
    response = requests.post(register, json={
        'url': json_endpoint.url,
        'secret': "Area 51",
        'method': method,
        'success_url': 'http://example.org',
        'error_url': 'http://example.org',
    })

    token = response.json()['token']

    response = requests.post(redirect, json={
        'foo': 'bar',
        'bar': token
    }, allow_redirects=False)

    assert response.status_code == 302

    result = json_endpoint.temporary_path / '{}.json'.format(method)
    with (result).open('r') as f:
        received = json.loads(f.read())

    assert received['foo'] == 'bar'
    assert received['bar'] == token
    assert received['oauth_redirect_secret'] == "Area 51"

    # simulate oauth provider that sends get
    response = requests.post(register, json={
        'url': json_endpoint.url,
        'secret': "Area 51",
        'method': method,
        'success_url': 'http://example.org',
        'error_url': 'http://example.org',
    })

    token = response.json()['token']

    response = requests.get(
        redirect + '?foo=bar&bar=' + token,
        allow_redirects=False
    )
    assert response.status_code == 302

    result = json_endpoint.temporary_path / '{}.json'.format(method)
    with (result).open('r') as f:
        received = json.loads(f.read())

    assert received['foo'] == 'bar'
    assert received['bar'] == token
    assert received['oauth_redirect_secret'] == "Area 51"


oauth_redirect/background.py
import asyncio

from datetime import datetime


async def start_tasks(app):
    app['remove_expired_tokens'] = app.loop.create_task(
        remove_expired_tokens(app))


async def cleanup_tasks(app):
    app['remove_expired_tokens'].cancel()
    await app['remove_expired_tokens']


async def remove_expired_tokens(app, periodicity=1.0):
    """ Regularly removes expired tokens from the database.

    Runs often because other code does not bother to check the validity
    of the token.

    """

    while True:
        horizon = datetime.utcnow()
        keys = []

        for key, token_bound in app.db.items():
            if token_bound['expiration'] < horizon:
                keys.append(key)

        for key in keys:
            del app.db[key]

        try:
            await asyncio.sleep(1.0)
        except asyncio.CancelledError:
            break


oauth_redirect/server.py
import json
import shelve

from aiohttp import ClientSession
from aiohttp import web
from aiohttp import web_exceptions as ex
from copy import copy
from datetime import datetime, timedelta
from itertools import chain
from json import JSONDecodeError
from oauth_redirect import background
from uuid import uuid4
from purl import URL


def is_authorized(request):
    """ Returns true if the given request has the configured auth code. """

    auth = request.match_info.get('auth')
    return auth is not None and auth == request.app.auth


def random_token():
    """ Generates random token. Uses a prefix in addition to an uuid value to
    make it easy to detect the value in a soup.

    """
    return '0x' + uuid4().hex


async def forward_request(token, request):
    """ Takes the given token and forwards the request to the associated
    redirect (must exist in database).

    Returns a web response with the same status code as the returned by
    the redirect.

    If the response was successful, the token is deleted immediately.

    """
    token_bound = request.app.db[token]
    url = token_bound['url']

    payload = copy(dict(request.query))
    payload.update(await request.json(loads=quiet_json))
    payload['oauth_redirect_secret'] = token_bound['secret']

    async with ClientSession() as session:

        if token_bound['method'] == 'GET':
            url = URL(url)

            for key, value in payload.items():
                url = url.query_param(key, value)

            def send_request():
                return session.get(url.as_string())
        elif token_bound['method'] == 'POST':
            def send_request():
                return session.post(url, data=json.dumps(payload))
        elif token_bound['method'] == 'PUT':
            def send_request():
                return session.put(url, data=json.dumps(payload))

        async with send_request() as response:

            if 200 <= response.status < 300:
                del request.app.db[token]
                return web.HTTPFound(token_bound['success_url'])
            else:
                return web.HTTPFound(token_bound['error_url'])


def quiet_json(data):
    """ Parses the given json string, ignoring any parsing errors and returning
    a dictionary in any case.

    """

    try:
        return json.loads(data)
    except JSONDecodeError:
        return {}


async def extract_token(request):
    """ Takes the given request and extracts the token from it. The token
    may be stored in any formdata, json data or query field.

    Whatever is first found among query, post and json data is presumed
    to be the token.

    """

    json_data = await request.json(loads=quiet_json)
    post_data = await request.post()

    values = chain(
        request.query.values(),
        json_data.values(),
        post_data.values(),
    )

    def is_token(value):
        return value.startswith('0x') and value in request.app.db

    return next((v for v in values if is_token(v)), None)


async def receive_redirect_to(request):
    """ Receives the OAuth redirect_to response and forwards it to a registerd
    client if a match can be found.

    """
    token = await extract_token(request)

    if token:
        return await forward_request(token, request)

    raise ex.HTTPNotFound(reason="Invalid token")


async def register(request):
    """ Registers a new redirect with the server. """

    if not is_authorized(request):
        raise ex.HTTPUnauthorized()

    data = await request.json(loads=quiet_json)
    ttl = int(data.get('ttl', 3600))

    token = random_token()

    try:
        method = data['method'].upper()

        if method not in ('GET', 'POST', 'PUT'):
            raise ex.HTTPBadRequest(reason="Unsupported HTTP method {}".format(
                method
            ))

        request.app.db[token] = {
            'url': data['url'],
            'expiration': datetime.utcnow() + timedelta(seconds=ttl),
            'secret': data['secret'],
            'method': method,
            'success_url': data['success_url'],
            'error_url': data['error_url']
        }
    except KeyError as e:
        raise ex.HTTPBadRequest(reason="Missing key: {}".format(e.args[0]))

    return web.json_response({'token': token})


def create_app(database, auth):
    """ Prepares the server application. """
    app = web.Application()

    app.db = shelve.open(database)
    app.auth = auth

    app.on_startup.append(background.start_tasks)
    app.on_cleanup.append(background.cleanup_tasks)

    app.router.add_get('/redirect', receive_redirect_to)
    app.router.add_post('/redirect', receive_redirect_to)
    app.router.add_post('/register/{auth}', register)

    return app


def run_server(host, port, database, auth):
    """ Prepares and runs the server application. """
    assert auth is not None

    web.run_app(host=host, port=port, app=create_app(database, auth))
